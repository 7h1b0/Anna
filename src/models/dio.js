const mongoose = require('mongoose');
const Joi = require('joi');

const Schema = mongoose.Schema;

const dio = new Schema({
  id_dio: { type: Number, required: true, unique: true },
  name: { type: String, required: true },
});

dio.statics.validate = function validate(data, callback) {
  const pattern = {
    id_dio: Joi.number()
      .integer()
      .required(),
    name: Joi.string()
      .trim()
      .min(3)
      .required(),
  };

  Joi.validate(data, pattern, callback);
};

dio.statics.pattern = function getPattern() {
  return {
    id_dio: Joi.number()
      .integer()
      .required(),
    name: Joi.string()
      .trim()
      .min(3)
      .required(),
  };
};

module.exports = mongoose.model('Dio', dio);
