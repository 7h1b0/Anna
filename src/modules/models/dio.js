import Ajv from 'ajv';
import dioSchema from '../schemas/dio';
import knex from '../../knexClient';
import { returnFirst } from '../dbUtil';

export const TABLE = 'dios';

export function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export function findAll() {
  return knex(TABLE).select({ dioId: 'dio_id' }, { roomId: 'room_id' }, 'name');
}

export function findById(id) {
  return returnFirst(
    knex(TABLE)
      .select({ dioId: 'dio_id' }, { roomId: 'room_id' }, 'name')
      .where('dio_id', id),
  );
}

export function save({ dioId, roomId, name }) {
  return knex(TABLE)
    .insert({ dio_id: dioId, room_id: roomId, name })
    .then(() => {
      return { dioId, roomId, name };
    });
}

export function remove(dioId) {
  return knex(TABLE)
    .where('dio_id', dioId)
    .del();
}

export function findByIdAndUpdate(dioId, payload) {
  return knex(TABLE)
    .update(payload)
    .where('dio_id', dioId);
}
