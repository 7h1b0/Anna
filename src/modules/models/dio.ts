import Ajv from 'ajv';
import dioSchema from '../schemas/dio';
import knex from '../../knexClient';

export const TABLE = 'dios';

export interface Dio {
  dioId: number;
  roomId: string;
  name: string;
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export async function findAll(): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name');
}

export async function findById(dioId: number): Promise<Dio> {
  return knex(TABLE)
    .first('dioId', 'roomId', 'name')
    .where('dioId', dioId);
}

export async function save(newDio: Dio): Promise<Dio> {
  return knex(TABLE)
    .insert(newDio)
    .then(() => newDio);
}

export function remove(dioId: number) {
  return knex(TABLE)
    .where('dioId', dioId)
    .del();
}

export function findByIdAndUpdate(dioId: number, payload: Partial<Dio>) {
  const { dioId: ignored, ...dio } = payload;
  return knex(TABLE)
    .update(dio)
    .where('dioId', dioId);
}
