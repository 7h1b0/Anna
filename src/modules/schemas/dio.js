module.exports = {
  $id: 'dio',
  type: 'object',
  additionalProperties: false,
  properties: {
    dioId: {
      type: 'integer',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    roomId: {
      type: 'integer',
    },
  },
  required: ['name', 'dioId', 'roomId'],
};
