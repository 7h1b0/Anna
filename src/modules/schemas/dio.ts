export default {
  $id: 'dio',
  type: 'object',
  additionalProperties: false,
  properties: {
    dioId: {
      type: 'integer',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    roomId: {
      type: 'string',
      format: 'uuid',
    },
  },
  required: ['name', 'dioId', 'roomId'],
};
