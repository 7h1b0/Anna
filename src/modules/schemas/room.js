export default {
  $id: 'room',
  type: 'object',
  additionalProperties: false,
  properties: {
    roomId: {
      type: 'string',
      format: 'uuid',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    description: {
      type: 'string',
      minLength: 3,
    },
  },
  required: ['name'],
};
