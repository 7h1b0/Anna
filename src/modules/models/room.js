import Ajv from 'ajv';
import uuidv4 from 'uuid/v4';
import roomSchema from '../schemas/room';
import knex from '../../knexClient';
import * as Dio from './dio';
import * as HueLight from './hueLight';
import { omit } from '../utils';

export const TABLE = 'rooms';
export const COLUMNS = [
  'roomId',
  'description',
  'name',
  'createdAt',
  'updatedAt',
  'createdBy',
];

export function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(roomSchema, data);
}

export function findAll() {
  return knex(TABLE).select(COLUMNS);
}

export function findById(id) {
  return knex(TABLE)
    .first(COLUMNS)
    .where('roomId', id);
}

export async function save({ name, description, userId }) {
  const roomId = uuidv4();
  await knex(TABLE).insert({
    roomId,
    description,
    name,
    createdBy: userId,
  });

  return roomId;
}

export async function remove(roomId) {
  const res = await knex
    .select('roomId')
    .from(Dio.TABLE)
    .where('roomId', roomId)
    .unionAll(function() {
      this.select('roomId')
        .from(HueLight.TABLE)
        .where('roomId', roomId);
    });

  if (res.length === 0) {
    return knex(TABLE)
      .where('roomId', roomId)
      .del();
  }
  return 0;
}

export function findByIdAndUpdate(roomId, payload) {
  const safePayload = omit(payload, [
    'roomId',
    'createdAt',
    'updatedAt',
    'createdBy',
  ]);
  return knex(TABLE)
    .update(safePayload)
    .where('roomId', roomId);
}
