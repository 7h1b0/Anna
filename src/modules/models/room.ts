import Ajv from 'ajv';
import uuidv4 from 'uuid/v4';
import roomSchema from '../schemas/room';
import knex from '../../knexClient';
import * as Dio from './dio';
import * as HueLight from './hueLight';

export const TABLE = 'rooms';
export const COLUMNS = [
  'roomId',
  'description',
  'name',
  'createdAt',
  'updatedAt',
  'createdBy',
];

export interface Room {
  roomId: string;
  name: string;
  description: string;
  createdAt: number;
  updatedAt: number;
  createdBy: string;
}

export function validate(data: any) {
  const ajv = new Ajv();
  return ajv.validate(roomSchema, data);
}

export async function findAll(): Promise<Room[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(id: string): Promise<Room> {
  return knex(TABLE)
    .first(COLUMNS)
    .where('roomId', id);
}

export async function save({
  name,
  description,
  userId,
}: {
  name: string;
  description: string;
  userId: string;
}): Promise<string> {
  const roomId = uuidv4();
  await knex(TABLE).insert({
    roomId,
    description,
    name,
    createdBy: userId,
  });

  return roomId;
}

export async function remove(roomId: string): Promise<number> {
  const res = await knex
    .select('roomId')
    .from(Dio.TABLE)
    .where('roomId', roomId)
    .unionAll(function() {
      this.select('roomId')
        .from(HueLight.TABLE)
        .where('roomId', roomId);
    });

  if (res.length === 0) {
    return knex(TABLE)
      .where('roomId', roomId)
      .del();
  }
  return 0;
}

export async function findByIdAndUpdate(
  roomId: string,
  payload: Partial<Room>,
): Promise<number> {
  const { roomId: ignored, ...room } = payload;
  return knex(TABLE)
    .update(room)
    .where('roomId', roomId);
}
