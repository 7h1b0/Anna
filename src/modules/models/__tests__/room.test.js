import knex from '../../../knexClient';
import * as Room from '../room';
import * as Dio from '../dio';
const initRooms = [
  {
    roomId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    description: 'this is a test',
    name: 'room_1',
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    roomId: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    description: 'this is a second test',
    name: 'room_2',
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
];

const dios = {
  dioId: 2,
  roomId: initRooms[1].roomId,
  name: 'test',
};

describe('Room', () => {
  beforeAll(async () => {
    await knex(Room.TABLE).truncate();
    await knex(Dio.TABLE).insert(dios);
  });

  beforeEach(async () => {
    await knex(Room.TABLE).insert(initRooms);
  });

  afterEach(async () => {
    await knex(Room.TABLE).truncate();
  });

  afterAll(async () => {
    await knex(Dio.TABLE).truncate();
    await knex.destroy();
  });

  describe('findAll', () => {
    it('should return all room', async () => {
      const result = await Room.findAll();
      expect(result).toEqual(initRooms);
    });
  });

  describe('findById', () => {
    it('should return only one room', async () => {
      const result = await Room.findById(initRooms[0].roomId);
      expect(result).toEqual(initRooms[0]);
    });

    it('should return undefined', async () => {
      const result = await Room.findById(-1);
      expect(result).toBeUndefined();
    });
  });

  describe('save', () => {
    it('should save a new room', async () => {
      const save = {
        name: 'test-save',
        description: 'test',
        userId: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
      };

      const newRoomId = await Room.save(save);
      const rooms = await knex(Room.TABLE)
        .first('*')
        .where('roomId', newRoomId);

      expect(rooms).toMatchSnapshot({
        roomId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      });
    });
  });

  describe('delete', () => {
    it('should delete a room', async () => {
      await Room.remove(initRooms[0].roomId);
      const rooms = await knex(Room.TABLE).select('*');
      expect(rooms).toEqual([initRooms[1]]);
    });

    it('should not delete a room', async () => {
      await Room.remove(-1);
      const rooms = await knex(Room.TABLE).select('*');
      expect(rooms).toEqual(initRooms);
    });

    it('should not delete a room if devices are still in it', async () => {
      const res = await Room.remove(initRooms[1].roomId);
      expect(res).toBe(0);
      const rooms = await knex(Room.TABLE).select('*');
      expect(rooms).toEqual(initRooms);
    });
  });

  describe('findByIdAndUpdate', () => {
    it('should update a room', async () => {
      const rowsAffected = await Room.findByIdAndUpdate(initRooms[0].roomId, {
        name: 'updated',
      });
      expect(rowsAffected).toBe(1);
      const room = await knex(Room.TABLE)
        .first('*')
        .where('roomId', initRooms[0].roomId);
      expect(room).toMatchSnapshot({
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      });
    });

    it('should not update a room', async () => {
      const rowsAffected = await Room.findByIdAndUpdate(-1, {
        name: 'updated',
      });
      expect(rowsAffected).toBe(0);
      const rooms = await knex(Room.TABLE).select('*');
      expect(rooms).toEqual(initRooms);
    });
  });

  describe('validate', () => {
    it('should return true when a room is valid', () => {
      const room = {
        name: 'room',
        description: 'test',
      };

      expect(Room.validate(room)).toBeTruthy();
    });

    it('should return false when a room is missing a props', () => {
      const room = {
        description: 'test',
      };

      expect(Room.validate(room)).toBeFalsy();
    });

    it('should return false when a room has unknow props', () => {
      const room = {
        name: 'room',
        description: 'test',
        test: 'test',
      };

      expect(Room.validate(room)).toBeFalsy();
    });
  });
});
