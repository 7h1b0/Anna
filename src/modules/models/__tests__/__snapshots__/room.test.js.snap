// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room findByIdAndUpdate should update a room 1`] = `
Object {
  "createdAt": Any<Date>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "description": "this is a test",
  "name": "updated",
  "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": Any<Date>,
}
`;

exports[`Room save should save a new room 1`] = `
Object {
  "createdAt": Any<Date>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "description": "test",
  "name": "test-save",
  "roomId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "updatedAt": Any<Date>,
}
`;
